library/core/src/alloc/mod.rs

```diff
@@ -87,7 +87,7 @@ impl fmt::Display for AllocError {
 /// # Safety
 ///
 /// * Memory blocks returned from an allocator must point to valid memory and retain their validity
-///   until the instance and all of its clones are dropped,
+///   until the instance and all of its clones become unreachable,
```

Comment:
This is, of course, the hard part of this PR.

I'm thinking of the guarantee in a similar way to that of references, in that it's a "while the handle exists." Of course, under the Stacked Borrows model (and, to an extent, NLL), it's _not_ when the borrow exists, but until when the borrow is last used.

So there're two ways I currently see to explain the weaker (not pin safe) guarantee, borrowing language from references:

- The memory regions are valid for the lifetime of the allocator. This is complicated by the written `'static` lifetime not actually being "the lifetime of the allocator" in this fashion, but an upper bound, so probably not a good way to explain it.
- The memory regions are valid until the point when the last clone is last used.

Effectively, what we're looking for is the dual of the pinning guarantee.

So I _think_ how I'd draft this is

> Memory blocks must retain their validity until the allocator instance and all of its clones are dropped, forgotten, leaked, or otherwise never accessed again. (In other words, when a lifetime held by the allocator instance is allowed to expire.)

(And now writing that I'm interested in how the `#[may_dangle]` borrowck eyepatch interacts with the guarantee, since that (very roughly) allows a lifetime to expire before the container. IIUC it doesn't, since the eyepatch refers to the lifetime within a generic argument, but I'm never certain w.r.t. may_dangle.)